### DISCUSSION TITLE
How to add links as a response to LLM model

### DISCUSSION THREAD

**Discussion by adityaworkfusion** [GoogleCloudPlatform/generative-ai#248] [General]
Hi,

Could you kindly help me with below need, and how it can be done.

**Need** - I want my LLM model to provide links along with the text to the response, is it possible that text-bison-001 model provides links to the question asked.

I am using chroma embeddings with LLMs, my dataset contains documentations texts. 

@polong-lin appreciate your time here.

Thanks
Aditya

**Comment by polong-lin** [comment]
One way to do this: when you lookup the embeddings, you should also lookup what the documentation source of that embedding is, then you can feed that link as part of the prompt to the LLM and do some prompt engineering to also return the documentation link alongside the response.

You can also look into [RetrievalQAWithSourcesChain from LangChain](https://api.python.langchain.com/en/latest/chains/langchain.chains.qa_with_sources.retrieval.RetrievalQAWithSourcesChain.html), and we have an [example of doing this with Vertex AI Search](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/search/retrieval-augmented-generation/examples/question_answering.ipynb):


``` 
from langchain.chains import RetrievalQAWithSourcesChain

retrieval_qa_with_sources = RetrievalQAWithSourcesChain.from_chain_type(
    llm=llm, chain_type="stuff", retriever=retriever
)
```

Output:
```
retrieval_qa_with_sources({"question": search_query}, return_only_outputs=True)
{'answer': 'DeepMind is a subsidiary of Alphabet. The CEO of DeepMind is Demis Hassabis.\n',
 'sources': '2020_alphabet_annual_report.pdf'}
```

**Reply by ghost** [reply]
how to implement for let's say i have TB's of documents on centrilized storage ex- sharepoint and i want to build a chatbot which should summerize user question and provide relevent link from sharepoint.
